libgraph/cirrus.c

```diff
@@ -0,0 +1,906 @@
+	{ GRAPH_640x480x24,  3, .gfx = { GRAPH_85Hz,  cfgs + 4 } },  /* 39 */
+	{ GRAPH_800x600x24,  3, .gfx = { GRAPH_60Hz,  cfgs + 5 } },  /* 40 */
+	{ GRAPH_800x600x24,  3, .gfx = { GRAPH_56Hz,  cfgs + 6 } },  /* 41 */
+	{ GRAPH_800x600x24,  3, .gfx = { GRAPH_72Hz,  cfgs + 7 } },  /* 42 */
+	{ GRAPH_800x600x24,  3, .gfx = { GRAPH_75Hz,  cfgs + 8 } },  /* 43 */
+	{ GRAPH_800x600x24,  3, .gfx = { GRAPH_85Hz,  cfgs + 9 } },  /* 44 */
+	{ GRAPH_1024x768x24, 3, .gfx = { GRAPH_60Hz,  cfgs + 10 } }, /* 45 */
+	{ GRAPH_1024x768x24, 3, .gfx = { GRAPH_43Hzi, cfgs + 11 } }, /* 46 */
+	{ GRAPH_1024x768x24, 3, .gfx = { GRAPH_70Hz,  cfgs + 12 } }, /* 47 */
+	{ GRAPH_1024x768x24, 3, .gfx = { GRAPH_75Hz,  cfgs + 13 } }, /* 48 */
+	{ GRAPH_1024x768x24, 3, .gfx = { GRAPH_85Hz,  cfgs + 14 } }, /* 49 */
+	/* 32-bit color (8:8:8:8) */
+	{ GRAPH_640x480x32,  4, .gfx = { GRAPH_60Hz,  cfgs + 1 } },  /* 50 */
+	{ GRAPH_640x480x32,  4, .gfx = { GRAPH_72Hz,  cfgs + 2 } },  /* 51 */
+	{ GRAPH_640x480x32,  4, .gfx = { GRAPH_75Hz,  cfgs + 3 } },  /* 52 */
+	{ GRAPH_640x480x32,  4, .gfx = { GRAPH_85Hz,  cfgs + 4 } },  /* 53 */
+	{ GRAPH_800x600x32,  4, .gfx = { GRAPH_60Hz,  cfgs + 5 } },  /* 54 */
+	{ GRAPH_800x600x32,  4, .gfx = { GRAPH_56Hz,  cfgs + 6 } },  /* 55 */
+	{ GRAPH_800x600x32,  4, .gfx = { GRAPH_72Hz,  cfgs + 7 } },  /* 56 */
+	{ GRAPH_800x600x32,  4, .gfx = { GRAPH_75Hz,  cfgs + 8 } },  /* 57 */
+	{ GRAPH_800x600x32,  4, .gfx = { GRAPH_85Hz,  cfgs + 9 } },  /* 58 */
+	/* No mode */
+	{ 0 }
+};
+
+
+/* VCLK table */
+static const cirrus_vclk_t vclks[] = {
+	/* Known stable VCLK */
+	{ 12599,  0x2c, 0x33 },
+	{ 18000,  0x27, 0x3e },
+	{ 19600,  0x28, 0x3a },
+	{ 25227,  0x4a, 0x2b },
+	{ 28325,  0x5b, 0x2f },
+	{ 31500,  0x42, 0x1f },
+	{ 36025,  0x4e, 0x3e },
+	{ 37747,  0x3a, 0x17 },
+	{ 39992,  0x51, 0x3a },
+	{ 41164,  0x45, 0x30 },
+	{ 45076,  0x55, 0x36 },
+	{ 49867,  0x65, 0x3a },
+	{ 64983,  0x76, 0x34 },
+	{ 72163,  0x7e, 0x32 },
+	{ 75000,  0x6e, 0x2a },
+	{ 80013,  0x5f, 0x22 },
+	{ 85226,  0x7d, 0x2a },
+	{ 89998,  0x58, 0x1c },
+	{ 95019,  0x49, 0x16 },
+	{ 100226, 0x46, 0x14 },
+	{ 108035, 0x53, 0x16 },
+	{ 109771, 0x5c, 0x18 },
+	{ 120050, 0x6d, 0x1a },
+	{ 125998, 0x58, 0x14 },
+	{ 130055, 0x6d, 0x18 },
+	{ 134998, 0x42, 0x0e },
+	{ 150339, 0x69, 0x14 },
+	{ 168236, 0x5e, 0x10 },
+	{ 188179, 0x5c, 0x0e },
+	{ 210679, 0x67, 0x0e },
+	{ 229088, 0x60, 0x0c },
+	/* No clock */
+	{ 0 }
+};
+
+
+/* Max VCLK for given color depth */
+static const unsigned int maxvclks[] = { 0, 135300, 86000, 86000, 60000 };
+
+
+/* Schedules and executes tasks */
+extern int graph_schedule(graph_t *graph);
+
+
+struct {
+	unsigned char init;
+} cirrus_common;
+
+
+int cirrus_vsync(graph_t *graph)
+{
+	cirrus_dev_t *cdev = (cirrus_dev_t *)graph->adapter;
+	void *ctx = cdev->ctx;
+
+	/* Wait for current vertical retrace end */
+	while (vgahw_status(ctx) & 0x08);
+	/* Wait for next vertical retrace start */
+	while (!(vgahw_status(ctx) & 0x08));
```

Comment:
You're right - vsync() implementation is the same for cirrus and vga graphics adapters. I thought about moving this function to libvga but it would produce name conflict (`vga_vsync(void *hwctx)` would be implemented in libvga and `vga_vsync(graph_t *graph)` would be implemented in libgraph/vga.c. Ofcourse the name could be changed (maybe `vga_waitvsync()` in libvga?) but I didn't bother for 2 lines of code.
Let me know if you think I should do that.

Reply:
If it's just two lines, no need to bother